File: packages/core/tests/compose-slot.test.ts
/**
 * Tests de `composeSlotValues` — KRO-73 (B+).
 *
 * Cubre formateo + filtering + orientation + separator + truncate.
 * El JSX que envuelve esto vive en Studio (`<ComposableSlot>`) y consume
 * el resultado.
 */

import { describe, it, expect } from 'vitest';
import { composeSlotValues, type ComposeSlotInput } from '../src/compose-slot';
import type { FieldDefLike } from '../src/types';

const slot = (overrides: Partial<ComposeSlotInput>): ComposeSlotInput => ({
  fields:      [],
  orientation: 'horizontal',
  separator:   ' · ',
  ...overrides,
});

const f = (value: unknown, behavior?: string): { def?: FieldDefLike; value: unknown } => ({
  def:   behavior ? { key: 'k', type: 'text', behavior } : { key: 'k', type: 'text' },
  value,
});

describe('composeSlotValues — items + filtering', () => {
  it('slot vacío → items vacío', () => {
    const r = composeSlotValues(slot({}));
    expect(r.items).toEqual([]);
  });

  it('values vacíos filtrados', () => {
    const r = composeSlotValues(slot({
      fields: [f('Madrid'), f(''), f(null), f('2024'), f(undefined)],
    }));
    expect(r.items).toEqual(['Madrid', '2024']);
  });

  it('aplica formatScalar por behavior', () => {
    const r = composeSlotValues(slot({
      fields: [f(4, 'rating'), f('2026-05-24', 'iso_date')],
    }));
    expect(r.items[0]).toBe('★★★★☆');
    expect(r.items[1]).toMatch(/2026/);
  });
});

describe('composeSlotValues — orientation / separator echo', () => {
  it('horizontal con separator custom → echo', () => {
    const r = composeSlotValues(slot({
      fields:      [f('a'), f('b')],
      orientation: 'horizontal',
      separator:   ' / ',
    }));
    expect(r.orientation).toBe('horizontal');
    expect(r.separator).toBe(' / ');
  });

  it('vertical → echo', () => {
    const r = composeSlotValues(slot({
      fields:      [f('línea 1'), f('línea 2')],
      orientation: 'vertical',
      separator:   ' · ',
    }));
    expect(r.orientation).toBe('vertical');
    expect(r.items).toEqual(['línea 1', 'línea 2']);
  });
});

describe('composeSlotValues — truncate', () => {
  it('sin charLimit → truncated null', () => {
    const r = composeSlotValues(slot({
      fields: [f('Madrid'), f('Barcelona')],
    }));
    expect(r.truncated).toBeNull();
  });

  it('charLimit=0 → no aplica', () => {
    const r = composeSlotValues(slot({
      fields:     [f('Madrid')],
      appearance: { truncateChars: 0 },
    }));
    expect(r.truncated).toBeNull();
  });

  it('texto MÁS corto que charLimit → no aplica', () => {
    const r = composeSlotValues(slot({
      fields:     [f('AB')],
      appearance: { truncateChars: 20 },
    }));
    expect(r.truncated).toBeNull();
  });

  it('texto que excede charLimit → truncated con "…"', () => {
    const r = composeSlotValues(slot({
      fields:     [f('Una frase razonablemente larga que excede')],
      appearance: { truncateChars: 10 },
    }));
    expect(r.truncated).toBe('Una frase…');
  });

  it('charLimit aplica al string COMPLETO joined (no por item)', () => {
    const r = composeSlotValues(slot({
      fields:     [f('aa'), f('bb'), f('cc')],
      separator:  ' · ',
      appearance: { truncateChars: 6 },
    }));
    // Joined: "aa · bb · cc" (12 chars). Cortado a 6 + "…".
    expect(r.truncated).toBeTruthy();
    expect(r.truncated!.endsWith('…')).toBe(true);
    expect(r.truncated!.length).toBeLessThanOrEqual(7);
  });

  it('orientation vertical usa joiner=" " para el truncate', () => {
    const r = composeSlotValues(slot({
      fields:      [f('línea1'), f('línea2')],
      orientation: 'vertical',
      separator:   ' · ', // ignorado en vertical
      appearance:  { truncateChars: 8 },
    }));
    // Joined: "línea1 línea2" (13 chars). Cortado a 8 + "…".
    expect(r.truncated).toBeTruthy();
    expect(r.truncated!.endsWith('…')).toBe(true);
  });
});

describe('composeSlotValues — sin def en field', () => {
  it('field sin def usa formatScalar fallback (string echo)', () => {
    const r = composeSlotValues(slot({
      fields: [{ value: 'plain' }],
    }));
    expect(r.items).toEqual(['plain']);
  });
});

File: packages/core_dart/test/compose_slot_test.dart
/// Corpus de `composeSlotValues` — ESPEJO 1:1 de `tests/compose-slot.test.ts`.
import 'package:test/test.dart';
import 'package:kromia_core/kromia_core.dart';

ComposeSlotInput slot({
  List<ComposeSlotField>? fields,
  String orientation = 'horizontal',
  String separator = ' · ',
  SlotAppearance? appearance,
}) =>
    ComposeSlotInput(
      fields: fields ?? const [],
      orientation: orientation,
      separator: separator,
      appearance: appearance,
    );

ComposeSlotField f(dynamic value, [String? behavior]) => ComposeSlotField(
      def: FieldDefLike(key: 'k', type: 'text', behavior: behavior),
      value: value,
    );

void main() {
  group('composeSlotValues — items + filtering', () {
    test('slot vacío → items vacío', () {
      expect(composeSlotValues(slot()).items, <String>[]);
    });

    test('values vacíos filtrados', () {
      final r = composeSlotValues(
          slot(fields: [f('Madrid'), f(''), f(null), f('2024'), f(null)]));
      expect(r.items, ['Madrid', '2024']);
    });

    test('aplica formatScalar por behavior', () {
      final r = composeSlotValues(
          slot(fields: [f(4, 'rating'), f('2026-05-24', 'iso_date')]));
      expect(r.items[0], '★★★★☆');
      expect(r.items[1], matches(RegExp(r'2026')));
    });
  });

  group('composeSlotValues — orientation / separator echo', () {
    test('horizontal con separator custom → echo', () {
      final r = composeSlotValues(
          slot(fields: [f('a'), f('b')], orientation: 'horizontal', separator: ' / '));
      expect(r.orientation, 'horizontal');
      expect(r.separator, ' / ');
    });

    test('vertical → echo', () {
      final r = composeSlotValues(
          slot(fields: [f('línea 1'), f('línea 2')], orientation: 'vertical'));
      expect(r.orientation, 'vertical');
      expect(r.items, ['línea 1', 'línea 2']);
    });
  });

  group('composeSlotValues — truncate', () {
    test('sin charLimit → null', () {
      expect(composeSlotValues(slot(fields: [f('Madrid'), f('Barcelona')])).truncated,
          isNull);
    });

    test('charLimit=0 → no aplica', () {
      expect(
          composeSlotValues(slot(
                  fields: [f('Madrid')],
                  appearance: const SlotAppearance(truncateChars: 0)))
              .truncated,
          isNull);
    });

    test('texto más corto que charLimit → no aplica', () {
      expect(
          composeSlotValues(slot(
                  fields: [f('AB')],
                  appearance: const SlotAppearance(truncateChars: 20)))
              .truncated,
          isNull);
    });

    test('texto que excede → truncated con …', () {
      final r = composeSlotValues(slot(
          fields: [f('Una frase razonablemente larga que excede')],
          appearance: const SlotAppearance(truncateChars: 10)));
      expect(r.truncated, 'Una frase…');
    });

    test('charLimit al string joined completo', () {
      final r = composeSlotValues(slot(
          fields: [f('aa'), f('bb'), f('cc')],
          separator: ' · ',
          appearance: const SlotAppearance(truncateChars: 6)));
      expect(r.truncated, isNotNull);
      expect(r.truncated!.endsWith('…'), isTrue);
      expect(r.truncated!.length, lessThanOrEqualTo(7));
    });

    test('vertical usa joiner espacio', () {
      final r = composeSlotValues(slot(
          fields: [f('línea1'), f('línea2')],
          orientation: 'vertical',
          separator: ' · ',
          appearance: const SlotAppearance(truncateChars: 8)));
      expect(r.truncated, isNotNull);
      expect(r.truncated!.endsWith('…'), isTrue);
    });
  });

  group('composeSlotValues — sin def', () {
    test('field sin def usa fallback string', () {
      final r = composeSlotValues(slot(fields: [const ComposeSlotField(value: 'plain')]));
      expect(r.items, ['plain']);
    });
  });
}
